### DISCUSSION TITLE
Custom Sampler doesn't have access to ASP.Net Core instrumented attributes

### DISCUSSION THREAD

**Discussion by AsakerMohd** [open-telemetry/opentelemetry-dotnet#5839] [Q&A] [answered]
We created a sampler that filters traces according to url.path and http.request.method. However, after doing some testing, we figured out that those attributes are not available (yet) when the sampling decision is made [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1c01770882bb5113ff308b2b4398347fab6e0404/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L246). FYI, we are using [opentelemetry auto instrumentation](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/)

Below is the stack trace showing the sampler working before any instrumentation takes place. 
```
at OpenTelemetry.Trace.TracerProviderSdk.RunGetRequestedDataOtherSampler(Activity activity)
   at OpenTelemetry.Trace.TracerProviderSdk.<>c__DisplayClass12_1.<.ctor>b__9(Activity activity)
   at System.Diagnostics.ActivitySource.<>c.<NotifyActivityStart>b__24_0(ActivityListener listener, Object obj)
   at System.Diagnostics.SynchronizedList`1.EnumWithAction(Action`2 action, Object arg)
   at System.Diagnostics.ActivitySource.NotifyActivityStart(Activity activity)
   at System.Diagnostics.Activity.Start()
``` 

Also, i saw [these comments](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L98) from ASP.Net Core instrumentation library mentioning that the samplers run before the `OnStartActivity` function. 

My question: Is there a way to overcome this so that our sampler has access to these attributes after instrumentation? In our sampler, we always set the sampling decision to RecordOnly instead of dropping it so that AllDataRequested field is true and populate span with attributes for metering purposes and Recorded to false to drop the trace so we will always instrument the trace with the required attributes. It's just that due to order of operation, these won't be available when evaluating the sampling decision. Is this by design? 

I know there were talks of tail sampling but that looks like it's still in proof of concept stage.

**Comment by cijothomas** [comment, marked as answer]
See similar/same discussion: https://cloud-native.slack.com/archives/C01N3BC2W7Q/p1725899641646069
